File: external/miscellaneous/ibm-cplex/cpoptimizer/examples/src/cpp/sched_truck_drone.cpp

// -------------------------------------------------------------- -*- C++ -*-
// File: ./examples/src/cpp/sched_truck_drone.cpp
// --------------------------------------------------------------------------
// Licensed Materials - Property of IBM
//
// 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5725-A06 5725-A29
// Copyright IBM Corporation 1990, 2024. All Rights Reserved.
//
// Note to U.S. Government Users Restricted Rights:
// Use, duplication or disclosure restricted by GSA ADP Schedule
// Contract with IBM Corp.
// --------------------------------------------------------------------------

/* ------------------------------------------------------------

Problem Description
-------------------

The Truck with Drone problem involves the collaboration between a truck and
a drone to deliver customers. The drone can pick up packages from the truck
and deliver them to the customer (but it cannot serve more than one location
at a time) while the truck is serving other customers.
The drone takes off or returns to the truck only at a customer location.
The problem consists in allocating delivery locations to either the truck
or the drone, while minimizing the total time to serve all customers, and
return to the depot.

------------------------------------------------------------ */

#include <ilcp/cp.h>

#define TIME_SCALE 1000000

IloInt IntegerTime(IloNum t) { return IloInt(t * TIME_SCALE); }
IloNum RealTime(IloInt t) { return t / (IloNum)TIME_SCALE; }
IloNumExpr RealTime(IloIntExpr t) { return t / (IloNum)TIME_SCALE; }


class ProblemData {
  IloNumArray _xco;
  IloNumArray _yco;
  IloNum      _truckSpeed;
  IloNum      _droneSpeed;
  IloInt      _numCustomers;

  void ignore(std::istream&in, IloInt n) {
    std::string s;
    for (IloInt i = 0; i < n; i++) in >> s;
  }

public:
  ProblemData(IloEnv env, std::istream& in);

  IloInt getNumCustomers() const { return _numCustomers; }
  IloInt getNumNodes() const { return _numCustomers + 1; }
  double getX(IloInt i) const { return _xco[i]; }
  double getY(IloInt i) const { return _yco[i]; }
  double getDistance(IloInt from, IloInt to) const {
    double dx = getX(from) - getX(to);
    double dy = getY(from) - getY(to);
    return IloPower(dx*dx + dy*dy, 0.5);
  }
  IloInt getDroneTime(IloInt from, IloInt to) const {
    return IntegerTime(getDistance(from, to) * _droneSpeed);
  }
  IloInt getTruckTime(IloInt from, IloInt to) const {
    return IntegerTime(getDistance(from, to) * _truckSpeed);
  }
};

ProblemData::ProblemData(IloEnv env, std::istream& in)
  : _xco(env)
  , _yco(env)
  , _truckSpeed(-1) 
  , _droneSpeed(-1) 
  , _numCustomers(-1) {
    ignore(in, 5); // ignore /*The speed of the Truck*/
    in >> _truckSpeed;
    ignore(in, 5); // ignore /*The speedo of the Drone*/
    in >> _droneSpeed;
    ignore(in, 3); // ignore /*Number of Nodes*/
    in >> _numCustomers;
    _numCustomers--;
    ignore(in, 2); // ignore /*The Depot*/
    IloNum x, y;
    in >> x >> y; ignore(in, 1); // ignore depot name
    _xco.add(x); _yco.add(y);
    ignore(in, 5); // ignore /*The Locations (x_coor y_coor name)*/
    for (IloInt i = 0; i < _numCustomers; i++) {
      in >> x >> y; ignore(in, 1); // ignore customer name
      _xco.add(x); _yco.add(y);
    }
}

const char * MakeName(IloEnv env, const char * base, IloInt i) {
  char * buf = new (env) char[32 + strlen(base)];
  snprintf(buf, 32 + strlen(base), "%s[%ld]", base, (long) i);
  return buf;
}

const char * MakeName(IloEnv env, const char * base, IloInt i, IloInt j) {
  char * buf = new (env) char[64 + strlen(base)];
  snprintf(buf, 64 + strlen(base), "%s[%ld][%ld]", base, (long)i, (long)j);
  return buf;
}


IloInt GetIndex(IloIntervalVarArray x, IloIntervalVar itv) {
  for (IloInt i = 0; i < x.getSize(); i++) {
    if (x[i].getImpl() == itv.getImpl()) {
      return i;
    }
  }
  return -1;
}

void GetBeforeAfter(IloCP cp,
                    IloInt dvIndex,
                    IloArray<IloIntervalVarArray> tdVisit,
                    IloArray<IloIntervalVarArray> dtVisit,
                    IloInt& dvBefore,
                    IloInt& dvAfter) {
  IloInt n = tdVisit.getSize();
  dvBefore = -1;
  dvAfter = -1;
  assert(dvIndex == -1 || (dvIndex >= 1 && dvIndex < n - 1));
  if (dvIndex >= 0) {
    for (IloInt j = 0; j < n; j++) {
      if (cp.isPresent(tdVisit[dvIndex][j])) {
        assert(dvBefore < 0);
        dvBefore = j;
      }
      if (cp.isPresent(dtVisit[dvIndex][j])) {
        assert(dvAfter < 0);
        dvAfter = j;
      }
    }
  }
  assert(dvIndex == -1 || dvBefore >= 0);
  assert(dvIndex == -1 || dvAfter >= 0);
}

void Solve(IloEnv env, const ProblemData& pd, IloNum tlim) {
  // Model variables
  IloInt n = pd.getNumNodes() + 1;
  IloIntervalVarArray    visit(env, n);
  IloIntervalVarArray    tVisit(env, n);
  IloIntervalVarArray    dVisit(env, n);
  IloIntervalVarArray    dVisitBefore(env, n);
  IloIntervalVarArray    dVisitAfter(env, n);
  IloIntArray            tVisitTypes(env, n);

  IloArray<IloIntervalVarArray> tdVisit(env, n);
  IloArray<IloIntervalVarArray> dtVisit(env, n);
  for (IloInt i = 0; i < n; i++) {
    visit[i] = IloIntervalVar(env, MakeName(env, "visit", i));

    tVisit[i] = IloIntervalVar(env, MakeName(env, "tVisit", i));
    tVisit[i].setOptional();

    dVisit[i] = IloIntervalVar(env, MakeName(env, "dVisit", i));
    dVisit[i].setOptional();

    dVisitBefore[i] = IloIntervalVar(env, MakeName(env, "dVisitBefore", i));
    dVisitBefore[i].setOptional();

    dVisitAfter[i] = IloIntervalVar(env, MakeName(env, "dVisitAfter", i));
    dVisitAfter[i].setOptional();

    tVisitTypes[i] = i;

    tdVisit[i] = IloIntervalVarArray(env, n);
    dtVisit[i] = IloIntervalVarArray(env, n);

    for (IloInt j = 0; j < n; j++) {
      tdVisit[i][j] = IloIntervalVar(env, MakeName(env, "tdVisit", i, j));
      tdVisit[i][j].setOptional();

      dtVisit[i][j] = IloIntervalVar(env, MakeName(env, "dtVisit", i, j));
      dtVisit[i][j].setOptional();

    }
  }
  tVisitTypes[n-1] = 0;
  IloIntervalSequenceVar tVisitSeq(env, tVisit, tVisitTypes);
  IloIntervalSequenceVar dVisitSeq(env, dVisit);

  IloArray<IloIntArray> truckTime(env, n - 1);
  for (IloInt i = 0; i < n - 1; i++) {
    truckTime[i] = IloIntArray(env, n - 1);
    for (IloInt j = 0; j < n - 1; j++)
      truckTime[i][j] = pd.getTruckTime(i, j);
  }

  // Constraints
  IloModel mdl(env);

  // Truck's depot visits must be present and bookend the customer visits
  tVisit[0].setPresent();
  tVisit[0].setStartMin(0);
  tVisit[0].setStartMax(0);
  tVisit[n-1].setPresent();
  mdl.add(IloFirst(env, tVisitSeq, tVisit[0]));
  mdl.add(IloLast(env, tVisitSeq, tVisit[n-1]));
  for (IloInt i = 1; i < n - 1; i++) {
    mdl.add(IloStartBeforeStart(env, tVisit[0], visit[i]));
    mdl.add(IloEndBeforeEnd(env, visit[i], tVisit[n-1]));
  }

  // Truck and drone can only do one thing at once.
  // Truck respects travel times.
  mdl.add(IloNoOverlap(env, tVisitSeq, truckTime));
  mdl.add(IloNoOverlap(env, dVisitSeq));

  for (IloInt i = 0; i < n; i++) {
    IloIntervalVarArray visitAlt(env);
    visitAlt.add(tVisit[i]); visitAlt.add(dVisit[i]);
    mdl.add(IloAlternative(env, visit[i], visitAlt));
    mdl.add(IloAlternative(env, dVisitBefore[i], tdVisit[i]));
    mdl.add(IloAlternative(env, dVisitAfter[i], dtVisit[i]));

    IloIntervalVarArray droneOutIn(env);
    droneOutIn.add(dVisitBefore[i]); droneOutIn.add(dVisitAfter[i]);
    mdl.add(IloSpan(env, dVisit[i], droneOutIn));
    mdl.add(IloEndAtStart(env, dVisitBefore[i], dVisitAfter[i]));
    mdl.add(IloPresenceOf(env, dVisit[i]) == IloPresenceOf(env, dVisitBefore[i]));
    mdl.add(IloPresenceOf(env, dVisit[i]) == IloPresenceOf(env, dVisitAfter[i]));
  }

  for (IloInt i = 1; i < n - 1; i++) { // Customer visits
    for (IloInt j = 0; j < n; j++) {
      IloInt outTime = pd.getDroneTime(tVisitTypes[j], tVisitTypes[i]);
      IloInt inTime = pd.getDroneTime(tVisitTypes[i], tVisitTypes[j]);
#if 1
      mdl.add(IloLengthOf(tdVisit[i][j], IloIntervalMax) >= outTime);
      mdl.add(IloLengthOf(dtVisit[i][j], IloIntervalMax) >= inTime);
#else
      tdVisit[i][j].setLengthMin(outTime); tdVisit[i][j].setLengthMax(outTime);
      mdl.add(IloLengthOf(dtVisit[i][j], IloIntervalMax) >= inTime);
#endif
      if (j >= 1 && j < n-1) {
        mdl.add(IloIfThen(env, IloPresenceOf(env, tdVisit[i][j]), IloPresenceOf(env, tVisit[j])));
        mdl.add(IloIfThen(env, IloPresenceOf(env, dtVisit[i][j]), IloPresenceOf(env, tVisit[j])));
      }
      mdl.add(IloStartBeforeStart(env, tVisit[j], tdVisit[i][j]));
      mdl.add(IloStartBeforeEnd(env, tdVisit[i][j], tVisit[j]));
      mdl.add(IloStartBeforeEnd(env, tVisit[j], dtVisit[i][j]));
      mdl.add(IloEndBeforeEnd(env, dtVisit[i][j], tVisit[j]));
    }
  }
  mdl.add(IloMinimize(env, RealTime(IloEndOf(visit[n-1]))));

  IloCP cp(mdl);
#if 0
  IloIntervalSequenceVarArray seqs(env); seqs.add(tVisitSeq); seqs.add(dVisitSeq);
  IloSearchPhaseArray phases(env);
  phases.add(IloSearchPhase(env, seqs));
  cp.setSearchPhases(phases);
#endif
  cp.setParameter(IloCP::TimeLimit, tlim);
  cp.setParameter(IloCP::LogPeriod, 100000);
  IloBool ok = cp.solve();

  cp.out() << std::endl;
  if (ok) {
    cp.out() << "Truck visits customers:";
    for (IloIntervalVar v = cp.getFirst(tVisitSeq);
         v.getImpl() != 0;
         v = cp.getNext(tVisitSeq, v)) {
      IloInt idx = GetIndex(tVisit, v);
      if (idx >= 1 && idx < n - 1)
        cp.out() << " " << idx;
    }
    cp.out() << std::endl;

    cp.out() << "Drone visits customers:";
    for (IloIntervalVar v = cp.getFirst(dVisitSeq);
         v.getImpl() != 0;
         v = cp.getNext(dVisitSeq, v)) {
      IloInt idx = GetIndex(dVisit, v);
      if (idx >= 1 && idx < n - 1)
        cp.out() << " " << idx;
    }
    cp.out() << std::endl;

    IloIntervalVar tv = cp.getFirst(tVisitSeq);
    IloIntervalVar dv = cp.getFirst(dVisitSeq);
    IloInt tvi = GetIndex(tVisit, tv);
    IloInt dvi = GetIndex(dVisit, dv);
    IloInt dvBefore = -1;
    IloInt dvAfter = -1;
    GetBeforeAfter(cp, dvi, tdVisit, dtVisit, dvBefore, dvAfter);
    cp.out() << std::endl;
    while (tvi >= 0) {
      // At loop entry, the drone is on the truck or joins it here.
      if (tvi >= 1 && tvi < n - 1)
        cp.out() << std::setw(3) << tvi;
      else
        cp.out() << "DEPOT";

      // Deal with drone flights while the truck does not move
      while (dvBefore == tvi && dvAfter == tvi) {
        cp.out() << " = " << dvi;
        dv = cp.getNext(dVisitSeq, dv);
        dvi = GetIndex(dVisit, dv);
        GetBeforeAfter(cp, dvi, tdVisit, dtVisit, dvBefore, dvAfter);
      }

      cp.out() << std::endl;
      const char * bb = "  | |"; // double bar
      const char * sb1 = "  |";  // single bar (truck)
      const char * sb2 = " |";   // double bar (drone)
      if (dvBefore == tvi) { // Is there a drone flight from here?
        // How many truck legs does the drone jump (1 means land next stop)
        IloInt jumps = 0;
        IloIntervalVar ntv = tv;
        while (ntv.getImpl() != tVisit[dvAfter].getImpl()) {
          ntv = cp.getNext(tVisitSeq, ntv);
          jumps++;
        }
        assert(jumps >= 1);
        cp.out() << sb1 << "\\" << std::endl;
        if (jumps % 2 == 1) { // odd number of jumps
          if (jumps > 1)
            cp.out() << bb << std::endl;
          for (IloInt i = 0; i < (jumps - 1) / 2; i++) {
            tv = cp.getNext(tVisitSeq, tv); tvi = GetIndex(tVisit, tv);
            cp.out() << bb << std::endl
                     << std::setw(3) << tvi << sb2 << std::endl
                     << bb << std::endl;
          }
          cp.out() << sb1 << " " << dvi << std::endl;
          for (IloInt i = 0; i < (jumps - 1) / 2; i++) {
            tv = cp.getNext(tVisitSeq, tv); tvi = GetIndex(tVisit, tv);
            cp.out() << bb << std::endl
                     << std::setw(3) << tvi << sb2 << std::endl
                     << bb << std::endl;
          }
          if (jumps > 1)
            cp.out() << bb << std::endl;
        }
        else { // even number of jumps
          for (IloInt i = 0; i < jumps / 2 - 1; i++) {
            tv = cp.getNext(tVisitSeq, tv); tvi = GetIndex(tVisit, tv);
            cp.out() << bb << std::endl << bb << std::endl
                     << std::setw(3) << tvi << sb2 << std::endl
                     << bb << std::endl;
          }
          tv = cp.getNext(tVisitSeq, tv); tvi = GetIndex(tVisit, tv);
          cp.out() << bb << std::endl << bb << std::endl
                   << std::setw(3) << tvi << " " << dvi << std::endl
                   << bb << std::endl << bb << std::endl;
          for (IloInt i = 0; i < jumps / 2 - 1; i++) {
            tv = cp.getNext(tVisitSeq, tv); tvi = GetIndex(tVisit, tv);
            cp.out() << bb << std::endl
                     << std::setw(3) << tvi << sb2 << std::endl
                     << bb << std::endl << bb << std::endl;
          }
        }
        cp.out() << sb1 << "/" << std::endl;
        dv = cp.getNext(dVisitSeq, dv); dvi = GetIndex(dVisit, dv);
        GetBeforeAfter(cp, dvi, tdVisit, dtVisit, dvBefore, dvAfter);
      }
      else if (tvi != n-1) { // no flight from here
        cp.out() << sb1 << std::endl << sb1 << std::endl << sb1 << std::endl;
      }
      tv = cp.getNext(tVisitSeq, tv); tvi = GetIndex(tVisit, tv);
    }
  }
  else {
    cp.out() << "No solution found" << std::endl;
  }
}

int main(int argc, const char * argv[]) {
  IloEnv env;
  try {
    const char * fname = "../../../examples/data/uniform-1-n11.txt";
    IloNum tlim = 10.0;
    if (argc > 1)
      fname = argv[1];
    if (argc > 2)
      tlim = atof(argv[2]);
    std::ifstream in(fname);
    if (!in.good())
      throw IloCP::Exception(0, "Could not open the data file");

    ProblemData pd(env, in);
    in.close();

    Solve(env, pd, tlim);
  }
  catch (IloException& ex) {
    env.out() << "Caught: " << ex << std::endl;
  }
  env.end();
  return 0;
}

File: external/miscellaneous/ibm-cplex/cpoptimizer/examples/src/java/SchedTruckDrone.java
// ------------------------------------------------------------- -*- Java -*-
// File: ./examples/src/java/SchedTruckDrone.java
// --------------------------------------------------------------------------
// Licensed Materials - Property of IBM
//
// 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5725-A06 5725-A29
// Copyright IBM Corporation 1990, 2024. All Rights Reserved.
//
// Note to U.S. Government Users Restricted Rights:
// Use, duplication or disclosure restricted by GSA ADP Schedule
// Contract with IBM Corp.
// --------------------------------------------------------------------------

/* ------------------------------------------------------------

Problem Description
-------------------

The TSP with Drone problem involves the collaboration between a truck and
a drone to deliver customers. The drone can pick up packages from the truck
and deliver them to the customer (but it cannot serve more than one location
at a time) while the truck is serving other customers.
The drone takes off or returns to the truck only at a customer location.
The problem consists in allocating delivery locations to either the truck
or the drone, while minimizing the total time to serve all customers, and
return to the depot.

The following model is an implementation of the paper "A Study on the Traveling
Salesman Problem with a Drone" (authors: Ziye Tang, Willem-Jan van Hoeve and Paul Shaw).
The sample data file "uniform-1-n11.txt" is also available at:
https://github.com/pcbouman-eur/TSP-D-Instances

------------------------------------------------------------ */

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StreamTokenizer;
import java.util.ArrayList;
import java.util.List;

import ilog.concert.IloException;
import ilog.concert.IloIntervalSequenceVar;
import ilog.concert.IloIntervalVar;
import ilog.concert.IloNumExpr;
import ilog.concert.IloTransitionDistance;
import ilog.cp.IloCP;

public class SchedTruckDrone {
    // Scaling factor for converting travel times to integers
    private static final int TIME_FACTOR = 1000000;

    private static class DataReader {

        private StreamTokenizer st;

        public DataReader(String filename) throws IOException {
            FileInputStream fstream = new FileInputStream(filename);
            Reader r = new BufferedReader(new InputStreamReader(fstream));
            st = new StreamTokenizer(r);
            st.wordChars('(', '_');  // Data file contains special characters that should be handled as regular "word" characters
        }

        public String next() throws IOException {
            st.nextToken();
            if (st.ttype == StreamTokenizer.TT_EOF)
                throw new IOException();
            return st.sval;
        }

        public int next_as_int() throws IOException {
            next();
            return (int) st.nval;
        }

        public double next_as_double() throws IOException {
            next();
            return st.nval;
        }

        public void skip_elems(int n) throws IOException {
            for (int i = 0; i < n; i++)
                next();
        }
    }

    /**
     * Holds a representation of the raw data of the problem.
     */
    private static class TSPDProblem {
        private double truckSpeed = -1;
        private double droneSpeed = -1;
        private int nbCustomers = -1;
        private List<double[]> xyList = new ArrayList<double[]>();
        private List<String> idsList = new ArrayList<String>();

        public TSPDProblem(String filename) throws IOException {
            read(filename);
        }

        public void read(String filename) throws IOException {
            DataReader data = new DataReader(filename);

            
            data.skip_elems(5); // /*The speed of the Truck*/
            this.truckSpeed = data.next_as_double();

            data.skip_elems(5); // /*The speed of the Drone*/
            this.droneSpeed = data.next_as_double();

            data.skip_elems(3); // /*Number of Nodes*/
            this.nbCustomers = data.next_as_int() - 1;

            data.skip_elems(2); // /*The Depot*/
            double[] locXy = new double[] {data.next_as_double(), data.next_as_double()};
            String locId = data.next();
            xyList.add( locXy );
            idsList.add( locId );

            data.skip_elems(5); /*The Locations (x_coor y_coor name)*/

            for (int cust_idx = 0; cust_idx < this.nbCustomers; cust_idx++) {
                locXy = new double[] {data.next_as_double(), data.next_as_double()};
                locId = data.next();
                xyList.add( locXy );
                idsList.add( locId );
            }
        }

        public int getNumNodes() { return nbCustomers + 1; }

        public double getTruckSpeed() { return truckSpeed; }

        public double getDroneSpeed() { return droneSpeed; }

        public double getDistance(int from_, int to_) {
            assert(from_ >= 0);
            assert(from_ <= getNumNodes());
            assert(to_ >= 0);
            assert(to_ <= getNumNodes());

            double[] c1 = xyList.get(from_);
            double[] c2 = xyList.get(to_);
            double dx = c2[0] - c1[0];
            double dy = c2[1] - c1[1];
            double d = Math.sqrt(dx * dx + dy * dy);
            return d;
        }

        public String getNodeId(int nodeIdx) {
            assert(nodeIdx >= 0);
            assert(nodeIdx <= getNumNodes());
            return idsList.get(nodeIdx);
        }
    }

    /**
     *  Represents the data structures of the model.
     */
    private static class TSPD extends IloCP {

        private TSPDProblem problem;
        
        private IloIntervalVar[] visit;
        private IloIntervalVar[] tVisit;
        private IloIntervalVar[] dVisit;
        private IloIntervalVar[] dVisitBefore;
        private IloIntervalVar[] dVisitAfter;
        private IloIntervalVar[][] tdVisit;
        private IloIntervalVar[][] dtVisit;
        
        private IloIntervalSequenceVar tVisitSeq;
        private IloIntervalSequenceVar dVisitSeq;

        public TSPD(TSPDProblem pb) throws IloException {
            initDataModel(pb);

            buildVariables();
            buildStructure();

            // Objective
            IloNumExpr obj = endOf(tVisit[problem.getNumNodes()]);
            add(minimize(obj));
        }

        private void buildVariables() throws IloException {
            int numCust = problem.getNumNodes() - 1;

            visit = new IloIntervalVar[numCust + 2];
            tVisit = new IloIntervalVar[numCust + 2];
            dVisit = new IloIntervalVar[numCust + 2];
            dVisitBefore = new IloIntervalVar[numCust + 2];
            dVisitAfter = new IloIntervalVar[numCust + 2];
            for (int i = 0; i < numCust + 2; i++) {
                visit[i] = intervalVar("visit_" + i);

                tVisit[i] = intervalVar("tVisit_" + i);
                tVisit[i].setOptional();

                dVisit[i] = intervalVar("dVisit_" + i);
                dVisit[i].setOptional();

                dVisitBefore[i] = intervalVar("dVisit_before_" + i);
                dVisitBefore[i].setOptional();

                dVisitAfter[i] = intervalVar("dVisit_after_" + i);
                dVisitAfter[i].setOptional();
            }

            int[] tVisitTypes = new int[numCust + 2];
            int[] dVisitTypes = new int[numCust + 2];
            for (int i = 0; i < numCust + 2; i++) {
                tVisitTypes[i] = i;
                dVisitTypes[i] = i;
            }
            tVisitTypes[numCust + 1] = 0;  // Last type corresponds to depot (index = 0)
            tVisitSeq = intervalSequenceVar(tVisit, tVisitTypes);
            dVisitSeq = intervalSequenceVar(dVisit, dVisitTypes);
    
            tdVisit = new IloIntervalVar[numCust + 2][numCust + 2];
            dtVisit = new IloIntervalVar[numCust + 2][numCust + 2];
            for (int i = 0; i < numCust + 2; i++) {
                for (int j = 0; j < numCust + 2; j++) {
                    tdVisit[i][j] = intervalVar("tdVisit_" + i + "_" + j);
                    tdVisit[i][j].setOptional();
    
                    dtVisit[i][j] = intervalVar("dtVisit_" + i + "_" + j);
                    dtVisit[i][j].setOptional();
                }
            }
        }

        private void buildStructure() throws IloException {
            int numCust = problem.getNumNodes() - 1;
        
            // Force presence of truck visits from and to depot
            tVisit[0].setPresent();
            tVisit[numCust + 1].setPresent();

            // Truck returning at depot is the last event
            for (int i = 0; i < numCust + 1; i++) {
                add(endBeforeStart(tVisit[i], tVisit[numCust + 1]));
            }
            add(first(tVisitSeq, tVisit[0]));
            add(last(tVisitSeq, tVisit[numCust + 1]));

            //
            int[][] distanceTable = new int[numCust + 1][numCust + 1];
            for (int i = 0; i < numCust + 1; i++) {
                for (int j = 0; j < numCust + 1; j++) {
                    distanceTable[i][j] = (int) (
                        problem.getDistance(i, j) * TIME_FACTOR * problem.getTruckSpeed()
                    );
                }
            }
            IloTransitionDistance distanceMatrix = transitionDistance(distanceTable);
            add(noOverlap(tVisitSeq, distanceMatrix));
            add(noOverlap(dVisitSeq));

            //
            for (int i = 0; i < numCust + 2; i++) {
                add(alternative(visit[i], new IloIntervalVar[] {tVisit[i], dVisit[i]}));
                add(alternative(dVisitBefore[i], tdVisit[i]));
                add(alternative(dVisitAfter[i], dtVisit[i]));

                add(span(dVisit[i], new IloIntervalVar[] {dVisitBefore[i], dVisitAfter[i]}));
                add(endAtStart(dVisitBefore[i], dVisitAfter[i]));
                add(ifThen(presenceOf(dVisit[i]), and(presenceOf(dVisitBefore[i]), presenceOf(dVisitAfter[i]))));
            }

            //
            for (int i = 0; i < numCust + 1; i++) {
                for (int j = 0; j < numCust + 2; j++) {
                    // Lower bounds on tdVisit and dtVisit
                    add(ge(lengthOf(tdVisit[i][j], IntervalMax), (int) (problem.getDistance((j < numCust + 1 ? j : 0), i) * TIME_FACTOR * problem.getDroneSpeed())));
                    add(ge(lengthOf(dtVisit[i][j], IntervalMax), (int) (problem.getDistance(i, (j < numCust + 1 ? j : 0)) * TIME_FACTOR * problem.getDroneSpeed())));

                    // Prevents warning about constraints that are always True
                    if ((j != 0) && (j != numCust + 1)) {
                        add(ifThen(presenceOf(tdVisit[i][j]), presenceOf(tVisit[j])));
                        add(ifThen(presenceOf(dtVisit[i][j]), presenceOf(tVisit[j])));
                    }
                    // Precedence constraints
                    add(startBeforeStart(tVisit[j], tdVisit[i][j]));
                    add(startBeforeEnd(tdVisit[i][j], tVisit[j]));
                    add(endBeforeEnd(dtVisit[i][j], tVisit[j]));
                    add(startBeforeEnd(tVisit[j], dtVisit[i][j]));   // ADDED to original model from Tang, van Hoeve & Shaw paper (otherwise may generate invalid solutions)
                }
            }
        }

        private void initDataModel(TSPDProblem pb) {
            this.problem = pb;
        }

        public double solve(double tlim) throws IloException {
            setParameter(IloCP.IntParam.SearchType, IloCP.ParameterValues.Auto);
            setParameter(IloCP.IntParam.LogPeriod, 1000000);
            setParameter(IloCP.DoubleParam.TimeLimit, tlim);

            exportModel("tspd_java.cpo");

            double objValue = Double.POSITIVE_INFINITY;
            if (solve()) {
                objValue = this.getObjValue();
            }
            return objValue;
        }
    }

    public static void main(String[] args) throws IOException {
        String filename = "../../../examples/data/uniform-1-n11.txt";
        double timeLimit = 5;
        if (args.length > 0)
            filename = args[0];
        if (args.length > 1)
            timeLimit = Double.valueOf(args[1]);

        TSPDProblem tspDProb = new TSPDProblem(filename);

        TSPD tspd = null;
        try {
            tspd = new TSPD(tspDProb);
            double obj = tspd.solve(timeLimit);
            if (obj != Double.POSITIVE_INFINITY)
                System.out.println("Found a solution with total time = " + (obj / TIME_FACTOR));
            else
                System.out.println("No solution found");
        }
        catch (IloException e) {
            System.err.println("Error " + e);
        }
        finally {
            if (tspd != null)
                tspd.end();
        }
    }
}
